sherpa/astro/io/__init__.py

```diff
@@ -948,6 +1066,463 @@ def convert(column, dtype):
+
+    Returns
+    -------
+    out : dict
+        The re-constructed data and header values.
+
+    Notes
+    -----
+    The F_CHAN, N_CHAN, and MATRIX columns are stored as either
+      - 1D arrays (unlikely for MATRIX); N_GRP = 1 for all rows
+      - 2D arrays when N_GRP is a constant for all rows
+      - 2D arrays when N_GRP <= 4 (so some rows will have 0's in
+        when N_GRP for that row is less than the max N_GRP)
+      - 2D array with dtype=object indicates a VLF (each row
+        should be the correct type though)
+
+    """
+
+    n_grp = []
+    f_chan = []
+    n_chan = []
+    matrix= []
+
+    # Used to reconstruct the original data
+    idx = 0
+    start = 0
+
+    # Header keyword
+    numelt = 0
+
+    # How big is the matrix?
+    matrix_size = set()
+
+    for ng in rmf.n_grp:
+        n_grp.append(ng)
+
+        f_chan.append([])
+        n_chan.append([])
+        matrix.append([])
+
+        # Short-cut when no data
+        if ng == 0:
+            matrix_size.add(0)
+
+            # Convert from [] to numpy empty list
+            f_chan[-1] = numpy.asarray([], dtype=numpy.int32)
+            n_chan[-1] = numpy.asarray([], dtype=numpy.int32)
+            matrix[-1] = numpy.asarray([], dtype=numpy.float32)
+            continue
+
+        # Grab the next ng elements from rmf.f_chan/n_chan
+        # and the corresponding rmf.matrix elements. Fortunately
+        # we can ignore F_CHAN when extracting data from matrix.
+        #
+        for _ in range(ng):
+            # ensure this is an integer and not a floating-point number
+            # which can happen when rmf.n_chan is stored as Unt64.
+            #
+            end = start + rmf.n_chan[idx].astype(numpy.int32)
+            mdata = rmf.matrix[start:end]
+
+            f_chan[-1].append(rmf.f_chan[idx])
+            n_chan[-1].append(rmf.n_chan[idx])
+            matrix[-1].extend(mdata)
+
+            idx += 1
+            start = end
+
+        # Ensure F_CHAN/N_CHAN is either 2- or 4-byte integer by
+        # converting to 4-byte integer here, which can later be
+        # downcast.
+        #
+        f_chan[-1] = numpy.asarray(f_chan[-1], dtype=numpy.int32)
+        n_chan[-1] = numpy.asarray(n_chan[-1], dtype=numpy.int32)
+
+        # Ensure the matrix is Real-4
+        matrix[-1] = numpy.asarray(matrix[-1], dtype=numpy.float32)
+        matrix_size.add(matrix[-1].size)
+
+        numelt += sum(n_chan[-1])
+
+    # N_GRP should be 2-byte integer.
+    #
+    n_grp = numpy.asarray(n_grp, dtype=numpy.int16)
+    numgrp = n_grp.sum()
+
+    # Can we convert F_CHAN/N_CHAN to fixed-length if either:
+    #  - N_GRP is the same for all rows
+    #  - max(N_GRP) < 4
+    #
+    if len(set(n_grp)) == 1 or n_grp.max() < 4:
+        ny = n_grp.max()
+        f_chan = _make_int_array(f_chan, ny)
+        n_chan = _make_int_array(n_chan, ny)
+    else:
+        f_chan = _make_int_vlf(f_chan)
+        n_chan = _make_int_vlf(n_chan)
+
```

Comment:
F_CHAN, N_CHAN are typically very small and by very small I mean that it's only a a few elements per row. Thus the space savings from a VLF are minimal, but the reading speed is slower (in any backend). So, once could argue that it makes sense to always have those as arrays of fixed length. This works and I'm not saying that we need ot change code the @DougBurke already wrote, I'm just suggesting that, if there are any edge cases that turn up later, we could just move it all to a fixed-length array.

Reply:
I'm basically trying to follow the advice given in the OGIP RMF description. We are likely to want to use VLF for the matrix array, so we need to handle VLF data anyway, but this is all internal routine so we can change it if there are found to be problems [ie we don't guarantee exactly how the serialization works, just that it should be "correct"].

This is just background information for the interested reader!